Add user tag to AWS and GCP volumes

james-boydell opened the pull request:
fixes: #1943 

-extended to gcp as well

james-boydell committed:
update volumes model

james-boydell committed:
generate db migration

james-boydell committed:
add user field to Volume class

james-boydell committed:
update volume service

- create volume now requires a user of type UserModel to be passed in
- db for volumes should have column for user_id
-only passing back the user name when creating the volume

james-boydell committed:
update router

james-boydell committed:
tag aws volume with dstack user that created it

james-boydell committed:
add datack user to gcp volume label

james-boydell committed:
updating tests

r4victor commented on src/dstack/_internal/server/routers/volumes.py:
```diff
@@ -66,12 +66,14 @@ async def get_volume(
+    user: UserModel = Depends(Authenticated()),
     user_project: Tuple[UserModel, ProjectModel] = Depends(ProjectMember()),
 ) -> Volume:
     _, project = user_project
```
No need to get user via Authenticated() since it's returned by ProjectMember(). Just do `user, project = user_project` like in other endpoints.

james-boydell replied:
fixed

r4victor commented on src/dstack/_internal/server/migrations/versions/9be642a33d08_added_user_id_and_user_to_volumes_table.py:
```diff
@@ -0,0 +1,45 @@
+            ondelete="CASCADE",
+        )
+
+    # ### end Alembic commands ###
```
This migration won't work on existing installations that already have volumes in the db. (You cannot add non-nullable field to existing rows).  I can suggest two solutions:

1. Just make VolumeModel.user_id/VolumeModel.user optional and work around that.
2. Ideally, we'd like to make user required. To do that, you can write a migration that adds a new nullable user field, then populates existing rows with values (e.g. from volume_model.project.owner), then makes the field non-nullable.

Solution 2 is preferable to avoid handling optional user case in the future, but I'm ok with merging Solution 1 as well if you're not feeling like writing data migrations.

james-boydell replied:
Good point. let me try and populate existing volumes with the user from volume's project owner.

james-boydell replied:
updated migration to set volume.user_id to the projects owner

r4victor commented:
@james-boydell, I also recommend running Python tests locally for a faster feedback loop. It should be fairly simple to set up: install deps from requirements_dev.txt and run `pytest src/tests/`.

james-boydell committed:
updating on how user is fetched

james-boydell committed:
updating tests

james-boydell committed:
update data migration to update volume user_id based on project.owner_id

james-boydell committed:
update volume model to remove lazy join

james-boydell committed:
regenerate migration

james-boydell committed:
join in users

r4victor committed:
Set default Volume.user for client backward compatibility

r4victor commented:
@james-boydell, thanks for the PR, nice work! I added a commit to handle backward compatibility. Merging it.

r4victor committed:
Load VolumeModel.user

The loading is required everywhere VolumeModel is converted to Volume.
If not loaded, the server errors when processing/returning other users' volumes.
E.g. when global admins lists all volumes.

r4victor commented:
Found missing VolumeModel.user loading in few places. Fixed it.

r4victor committed:
Fix VolumeModel.user loading on postgres
